### PR TITLE
Add Zig CUDA GLiNER2.5 inference and fine-tuning

### PR DESCRIPTION
**Pull request by timkaye11**
# Add Zig CUDA GLiNER2.5 inference and FP32 fine-tuning

## Summary

Adds complete GLiNER2.5 CUDA inference and FP32 fine-tuning support to the Zig
backend, reusing the shared CPU/Metal graphs, tensor inventory, objectives,
fixtures, checkpoint contracts, and benchmark harnesses.

## What changed

- Added resident CUDA GLiNER2.5 forward/backward graphs and training plans.
- Added CUDA kernels for boundary attention, encoder execution, span features,
  reductions, normalization, softmax, losses, gradient accumulation, clipping,
  AdamW, checkpoint resume, and portable export.
- Added deterministic generated CUDA artifacts with provenance manifests and
  regeneration tests.
- Added explicit native-training cuBLAS selection through
  `ANTFLY_INFERENCE_CUDA_TRAINING_CUBLAS_LIBRARY`, including symbol/version
  validation and fail-closed behavior.
- Added cuBLAS identity to trainer readiness and checkpoint fingerprints so
  incompatible resumes are rejected safely.
- Extended the Python CUDA worker and benchmark harness for matching cuBLAS
  identity, parity checks, allocation diagnostics, and encoder admission.
- Preserved strict CUDA memory limits, transfer accounting, cancellation,
  rollback, snapshot completeness, and host/RSS supervision.

## Validation

All sustained comparisons use deterministic float32 fixtures, dropout disabled,
accumulation two, Adam epsilon `1e-8`, 100 updates, and cuBLAS 12.8 on both arms.
Weights, gradients, and Adam moments pass scheduled and final parity checks.

| Configuration | Tensors | Zig ex/s | Python CUDA ex/s | Relative speed |
| --- | ---: | ---: | ---: | ---: |
| Small, heads, B8 | 136 | 90.16 | 52.75 | Zig 1.71x faster |
| Small, full, B2 | 334 | 18.77 | 9.03 | Zig 2.08x faster |
| Small, full, B8 | 334 | 40.01 | 33.84 | Zig 1.18x faster |
| Multilingual, heads, B8 | 136 | 74.70 | 52.22 | Zig 1.43x faster |

All four rows pass numerical parity and favor Zig. Python contract tests pass
23/23. Focused Compute Sanitizer memcheck passes the selected CUDA training,
clipping, and seeded-trainer cases with zero errors. The CUDA test and trainer
builds completed; a later full inference rebuild reached linking but could not
write its executable after the local filesystem filled.

## Runtime compatibility

To reproduce parity with the pinned PyTorch CUDA runtime, set:

```text
ANTFLY_INFERENCE_CUDA_TRAINING_CUBLAS_LIBRARY=/absolute/path/to/libcublas.so.12
```

Serving cuBLASLt selection is unchanged. Checkpoint resumes require matching
recorded CUDA/cuBLAS runtime identity; portable weight exports remain loadable.

## Remaining qualification

Before production registry enablement, complete the frozen real-data matrix,
base/multilingual full-head coverage, LoRA/DoRA adapter fixtures,
`torch.compile`/FlashDeBERTa baseline selection, and Ampere/Hopper validation.
BF16/FP16 training, CUDA graphs, persistent workspaces, and additional cuBLASLt
tuning remain deferred.

## Reviewer checklist

- [ ] Review shared graph/objective changes for CPU and Metal compatibility.
- [ ] Review CUDA ownership, admission, rollback, and cleanup paths.
- [ ] Verify generated artifact provenance and regeneration commands.
- [ ] Re-run focused Zig, Python contract, and Compute Sanitizer checks.
- [ ] Confirm deployment cuBLAS path/hash and checkpoint fingerprint policy.

### CODE REVIEW

File: zig/pkg/inference/src/architectures/gliner_boundary_graph.zig
Comment (ajroetker) [inline zig/pkg/inference/src/architectures/gliner_boundary_graph.zig:9]: preexisting but `zig/pkg/inference/src/architectures/gliner/boundary_graph.zig` gliner might deserve it's own architecture module? instead of prefixed files?

File: zig/pkg/inference/src/finetune/gliner_boundary_native_trainer.zig [hunk 1]
```diff
@@ -50,7 +50,10 @@ pub const Limits = struct {
     peft: peft.Limits = .{},
     differentiation: seeded.Options = .{},
 };
+pub const DecisionEvents = @import("gliner_boundary_training_decisions.zig");
```
Comment (ajroetker) [inline zig/pkg/inference/src/finetune/gliner_boundary_native_trainer.zig]: same here but `zig/pkg/inference/src/finetune/gliner/boundary_native_trainer.zig` maybe we move to a gliner module/dir within because the file list is huge?

Comment (ajroetker) [reply zig/pkg/inference/src/finetune/gliner_boundary_native_trainer.zig]: unless there's a reason not to

**Comment by timkaye11** [comment]
looks like there were no gliner2.5 or gliner2 related failures in the latest CI run, and GPU/inference jobs passed. Failures were 
  - x86 Zig unit suite: watchdog/hang
  - base Antfly E2E: restore/metadata path
